proposals/rich-content-graphql-api.md

```diff
@@ -0,0 +1,331 @@
+        "preload": ["banner1.jpg"],
+        "descendentRoles": []
+      },
+      {
+        "role": "Accordion",
+        "depth": 3,
+        "preload": [
+          "diagram.svg",
+          "thumbnail001.jpg",
+          "thumbnail002.jpg",
+          "video.mp4"
+        ],
+        "descendentRoles": ["Video", "Image", "Row"]
+      },
+      {
+        "role": "Text",
+        "depth": 0,
+        "preload": [],
+        "descendentRoles": []
+      }
+    ],
+    "html": "Hello world! Text has already loaded."
+  }
+}
+```
+
+The PWA would handle this response:
+
+1.  In parallel:
+
+    1.  Render the `html` content by injecting it into a `RawHTML` component.
+
+    2.  Initiate preload requests for all the URLs in the top-level `preload` field. *Note that the responses "roll up" descendent preloads to a topmost list, as well as descendentRoles.*
+
+    3.  In series:
+
+        1.  In parallel, dynamically load and inject the React Components for:
+
+            - `PageBuilder_Column`
+            - `PageBuilder_Banner`
+            - `PageBuilder_Accordion`
+            - `PageBuilder_Video`
+            - `PageBuilder_Image`
+            - `PageBuilder_Row`
+            - `PageBuilder_Text`
+
+            (The bundle system and service worker will load from their caches if possible.)
+
+        2.  Run a subsequent query provided by the components' respective implementations:
+
+            ```graphql
+            fragment PageBuilder_Column_Config on PageBuilder_Column {
+              appearance
+              backgroundColor
+              backgroundImage
+              backgroundPosition
+              backgroundSize
+              backgroundRepeat
+              backgroundAttachment
+              minHeight
+              verticalAlign
+            }
+
+            fragment PageBuilder_Banner_Config on PageBuilder_Banner {
+              appearance
+              backgroundType
+              "More banner config here."
+            }
+
+            """
+            Additional fragments here.
+            """
+
+            content {
+              ...PageBuilder_Column_Config
+              children {
+                ... on PageBuilder_Banner {
+                  ...PageBuilder_Banner_Config
+                }
+                ... on PageBuilder_Accordion {
+            	  ...PageBuilder_Accordion_Config
+                }
+                ... on PageBuilder_Text {
+                  ...PageBuilder_Text_Config
+                }
+            	children {
+                  ...NodeMetadata
+                }
+              }
+            }
+            ```
+
+        3.  Re-render the fully hydrated second level of component when that query returns results.
+
+Subsequent requests for any content which lists any of those components in its `descendentRoles` will automatically include those fragment and clauses in the `content` query. Therefore, subsequent requests will contain the necessary configurations on the first pass.
```

Comment:
Would be good to measure performance on the prototype. I feel like network communication expenses would be higher than potential slowness of getting large JSON in one request. I think we wouldn't have a lot of content. This also makes caching less efficient.

Reply:
We should definitely measure the performance, and change approaches if it's terrible.

I think that caching might be better with GraphQL. If VisualContentNodes have unique IDs, then [Apollo cache implementations](https://github.com/convoyinc/apollo-cache-hermes) can store their representations incrementally. Then, the PWA would reuse references to the same content entity.